File: _core/lib/sw2/edit-mons.js
"use strict";
const gameSystem = 'sw2';

window.onload = function() {
  setName();
  rewriteMountLevel();
  updatePartsAutomatically();
  updatePartList();
  selectInputCheck(form.taxa,'その他')
  checkMount();

  changeColor();
}

// 送信前チェック ----------------------------------------
function formCheck(){
  if(form.monsterName.value === '' && form.characterName.value === ''){
    alert('名称か名前のいずれかを入力してください。');
    form.monsterName.focus();
    return false;
  }
  if(form.protect.value === 'password' && form.pass.value === ''){
    alert('パスワードが入力されていません。');
    form.pass.focus();
    return false;
  }
  return true;
}

// 名前 ----------------------------------------
function setName(){
  let m = ruby(form.monsterName.value);
  let c = ruby(form.characterName.value);
  document.querySelector('#header-menu > h2 > span').innerHTML = c && m ? `${c}<small>（${m}）</small>` : (c || m || '(名称未入力)');

  function vCheck(id){
    if(form[id]){ return form[id].value; }
    else { return '' }
  }
}
// 騎獣 ----------------------------------------
let mountFlag = 0;
function checkMount(){
  mountFlag = form.mount.checked ? 1 : 0;
  form.classList.toggle('mount', mountFlag);
}
function checkLevel(){
  if(mountFlag){
    checkMountLevel();
  }
}
// 各ステータス計算 ----------------------------------------
function calcVit(){
  const val = form.vitResist.value;
  form.vitResistFix.value = (val == '') ? '' : Number(val) + 7;
}
function calcVitF(){
  const val = form.vitResistFix.value;
  form.vitResist.value    = (val == '') ? '' : Number(val) - 7;
}
function calcMnd(){
  const val = form.mndResist.value;
  form.mndResistFix.value = (val == '') ? '' : Number(val) + 7;
}
function calcMndF(){
  const val = form.mndResistFix.value;
  form.mndResist.value    = (val == '') ? '' : Number(val) - 7;
}
function calcAcc(Num){
  const val = form['status'+Num+'Accuracy'].value;
  form['status'+Num+'AccuracyFix'].value = (val == '') ? '' : Number(val) + 7;
}
function calcAccF(Num){
  const val = form['status'+Num+'AccuracyFix'].value;
  form['status'+Num+'Accuracy'].value    = (val == '') ? '' : Number(val) - 7;
}
function calcEva(Num){
  const val = form['status'+Num+'Evasion'].value;
  form['status'+Num+'EvasionFix'].value  = (val == '') ? '' : Number(val) + 7;
}
function calcEvaF(Num){
  const val = form['status'+Num+'EvasionFix'].value;
  form['status'+Num+'Evasion'].value     = (val == '') ? '' : Number(val) - 7;
}

// ステータス欄 ----------------------------------------
function checkMountLevel(){
  let min = Number(form.lvMin.value) || 0;
  let max = Number(form.lvMax.value) || 0;
  if(max < min){ form.lvMax.value = max = min }
  if(form.lv.value != ''){
    if(form.lv.value < min){ form.lv.value = min }
    if(form.lv.value > max){ form.lv.value = max }
  }
  let gap = max - min;
  gap = gap < 0 ? 0 : gap;
  if(gap > 0){
    for(let lv = 2; lv <= gap+1; lv++){
      if(!document.getElementById(`status-tbody${lv}`)){
        let tbody = document.createElement("tbody");
        tbody.classList.add('mount-only');
        tbody.id = `status-tbody${lv}`;
        tbody.dataset.lv = lv;
        document.getElementById('status-table').append(tbody);
        for(let num = 1; num <= form.statusNum.value; num++){
          addStatusInsert(tbody, num);
        }
      }
    }
  }
  for(let lv = gap+2; document.getElementById(`status-tbody${lv}`); lv++){
    document.getElementById(`status-tbody${lv}`).remove();
  }
  for(let num = 1; num <= form.statusNum.value; num++){ checkStyle(num); }
  rewriteMountLevel(min);
}
function rewriteMountLevel(level){
  level ||= form.lvMin.value;
  document.querySelectorAll("#status-table tbody tr th:first-child").forEach(obj => {
    obj.textContent = '';
  });
  document.querySelectorAll("#status-table tbody tr:first-child th:first-child").forEach(obj => {
    obj.textContent = level;
    obj.classList.toggle('current', level == form.lv.value);
    level++;
  });
}
// 攻撃方法
function checkStyle(num){
  document.querySelectorAll(`#status-table .name[data-style="${num}"]`).forEach(obj => {
    obj.textContent = form[`status${num}Style`].value;
  });
}
// 追加・複製
function addStatus(copy){
  let num = Number(form.statusNum.value) + 1;
  document.querySelectorAll("#status-table tbody").forEach(obj => {
    addStatusInsert(obj, num, copy);
  });
  form.statusNum.value = num;
  statusTextInputToggle();
  updatePartsAutomatically();
}
function addStatusInsert(target, num, copy){
  const lv = target.dataset.lv ? '-'+target.dataset.lv : '';
  const ini = {
    "style"      : copy && !lv ? form[`status${copy}${lv}Style`       ].value : '',
    "accuracy"   : copy        ? form[`status${copy}${lv}Accuracy`    ].value : '',
    "accuracyFix": copy && !lv ? form[`status${copy}${lv}AccuracyFix` ].value : '',
    "damage"     : copy        ? form[`status${copy}${lv}Damage`      ].value : '2d+',
    "evasion"    : copy        ? form[`status${copy}${lv}Evasion`     ].value : '',
    "evasionFix" : copy && !lv ? form[`status${copy}${lv}EvasionFix`  ].value : '',
    "defense"    : copy        ? form[`status${copy}${lv}Defense`     ].value : '',
    "hp"         : copy        ? form[`status${copy}${lv}Hp`          ].value : '',
    "mp"         : copy        ? form[`status${copy}${lv}Mp`          ].value : '',
    "vit"        : copy        ? form[`status${copy}${lv}Vit`         ].value : (num == 1 ? '' : '―'),
    "mnd"        : copy        ? form[`status${copy}${lv}Mnd`         ].value : (num == 1 ? '' : '―'),
  };
  let tr = document.createElement('tr');
  tr.setAttribute('id',idNumSet('status-row',lv));
  tr.innerHTML = `
    <th class="mount-only"></th>
    <td ${ lv ? '' : `class="handle"`}></td>
    <td ${ lv ? 'class="name"' : ``} data-style="${num}">${ lv ? form[`status${num}Style`].value : `<input name="status${num}${lv}Style" type="text" value="${ini.style}" oninput="checkStyle(${num}${lv}); updatePartsAutomatically();">` }</td>
    <td>
      <input name="status${num}${lv}Accuracy" type="text" oninput="calcAcc('${num}${lv}')" value="${ini.accuracy}"><span class="monster-only calc-only"><br>
      (<input name="status${num}${lv}AccuracyFix" type="text" oninput="calcAccF('${num}${lv}')" value="${ini.accuracyFix}">)</span>
    </td>
    <td><input name="status${num}${lv}Damage" type="text" value="${ini.damage}"></td>
    <td>
      <input name="status${num}${lv}Evasion" type="text" oninput="calcEva('${num}${lv}')" value="${ini.evasion}"><span class="monster-only calc-only"><br>
      (<input name="status${num}${lv}EvasionFix" type="text" oninput="calcEvaF('${num}${lv}')" value="${ini.evasionFix}">)</span>
    </td>
    <td><input name="status${num}${lv}Defense" type="text" value="${ini.defense}"></td>
    <td><input name="status${num}${lv}Hp" type="text" value="${ini.hp}"></td>
    <td><input name="status${num}${lv}Mp" type="text" value="${ini.mp}"></td>
    <td class="mount-only"><input name="status${num}${lv}Vit" type="text" value="${ini.vit}"></td>
    <td class="mount-only"><input name="status${num}${lv}Mnd" type="text" value="${ini.mnd}"></td>
    <td>${ lv ? '' : `<span class="button" onclick="addStatus('${num}${lv}');">複<br>製</span>` }</td>
  `;
  target.appendChild(tr, target);
}
// 削除
function delStatus(){
  let num = Number(form.statusNum.value);
  if(num > 1){
    let hasValue = false;
    for (const node of document.querySelectorAll(`#status-table tbody tr:last-child input`)){
      if(
        node.value !== '' &&
        !(/Damage$/.test(node.getAttribute('name')) && node.value === '2d+') &&
        !(/Vit$/.test(node.getAttribute('name')) && node.value === '―') &&
        !(/Mnd$/.test(node.getAttribute('name')) && node.value === '―')
      ){
        hasValue = true; break;
      }
    }
    if(hasValue){
      if (!confirm(delConfirmText)){ return false; }
    }
    document.querySelectorAll("#status-table tbody tr:last-child").forEach(target => {
      target.remove();
    });
    num--;
    form.statusNum.value = num;
  }
  updatePartsAutomatically();
}
// ソート
(() => {
  let sortable = Sortable.create(document.querySelector('#status-table tbody'), {
    dataIdAttr: 'id',
    animation: 150,
    handle: '.handle',
    filter: 'thead,tfoot',
    onUpdate: function (evt) {
      const order = sortable.toArray();
      let num = 1;
      for(let id of order) {
        const row = document.querySelector(`tr#${id}`);
        if(!row) continue;
        row.querySelectorAll('[name]').forEach(inputField => {
          const beforeName = inputField.getAttribute('name');
          const afterName = beforeName.replace(/^(status)\d+(.+)$/, `$1${num}$2`);
          inputField.setAttribute('name', afterName)
        });
        row.querySelectorAll('[oninput]').forEach(inputField => {
          const beforeName = inputField.getAttribute('oninput');
          const afterName = beforeName.replace(/\(\d+\)/, `(${num})`);
          inputField.setAttribute('oninput', afterName)
        });
        row.querySelector(`span[onclick]`).setAttribute('onclick',`addStatus(${num})`);
        num++;
      }
      const moved  = evt.item.id;
      const before = evt.item.previousElementSibling ? evt.item.previousElementSibling.id : '';
      document.querySelectorAll("#status-table tbody").forEach(obj => {
        const lv = obj.dataset.lv;
        if(lv){
          if(before){
            document.getElementById(before+'-'+lv).after(document.getElementById(moved+'-'+lv));
          }
          else {
            document.getElementById(`status-tbody${lv}`).prepend(document.getElementById(moved+'-'+lv))
          }
          let num = 1;
          for(let id of order) {
            const row = document.querySelector(`tr#${id}-${lv}`);
            if(!row) continue;
            row.querySelectorAll('[name]').forEach(inputField => {
              const beforeName = inputField.getAttribute('name');
              const afterName = beforeName.replace(/^(status)\d+-(.+)$/, `$1${num}-$2`);
              inputField.setAttribute('name', afterName)
            });
            row.querySelector(`.name`).dataset.style = num;
            num++;
          }
        }
      });
      rewriteMountLevel();
      updatePartsAutomatically();
    }
  });
})();
//
function statusTextInputToggle(){
  const on = form.statusTextInput.checked ? 1 : form.mount.checked ? 1 : 0;
  form[`vitResist`].type    = on ? 'text'   : 'number';
  form[`mndResist`].type    = on ? 'text'   : 'number';
  for(let i = 1; i <= form.statusNum.value; i++){
    form[`status${i}Accuracy`].type    = on ? 'text'   : 'number';
    form[`status${i}Evasion`].type     = on ? 'text'   : 'number';
  }
  form.classList.toggle('not-calc', on)
}
// 部位数・内訳の自動入力
function updatePartsAutomatically() {
  const manualModeCheckbox = document.querySelector('input[type="checkbox"][name="partsManualInput"]');
  const partsNumInput = document.querySelector('.parts input[name="partsNum"]');
  const partsNamesInput = document.querySelector('.parts input[name="parts"]');

  if (manualModeCheckbox.checked) {
    partsNumInput.readOnly = false;
    partsNamesInput.readOnly = false;
    return;
  }

  let partCount = 0;
  const partNames = [];
  document.querySelectorAll('#status-tbody input[name$="Style"]').forEach(
      input => {
        partCount++;

        const style = input.value.trim();
        const m = style.match(/.*[(（](.+?)[）)]$/);
        if (m == null) {
          return;
        }
        partNames.push(m[1].trim());
      }
  );

  partsNumInput.readOnly = true;
  partsNumInput.value = partCount.toString();
  partsNumInput.dispatchEvent(new Event('input'));

  partsNamesInput.readOnly = true;
  partsNamesInput.value = partNames.length === 0 ? '' : partNames.reduce(
      (previous, currentPartName) => {
        const previousPartTexts = previous.split('／');
        const lastPartText = previousPartTexts[previousPartTexts.length - 1];
        const m = lastPartText.match(/^(.+?)(?:×(\d+))?$/);
        const lastPartName = m[1];
        const lastPartCount = m[2] ? parseInt(m[2]) : 1;
        return currentPartName === lastPartName
            ? `${previousPartTexts.length > 1 ? `${previousPartTexts.slice(0, -1).join('／')}／` : ''}${lastPartName}×${lastPartCount + 1}`
            : `${previous}／${currentPartName}`;
      }
  );
  partsNamesInput.dispatchEvent(new Event('input'));
}
function updatePartList() {
  const partsText = document.querySelector('input[name="parts"]').value.trim();

  const items =
      partsText
          .split(/[/／]/)
          .map(x => x.trim())
          .filter(x => x !== '')
          .map(part => part.replace(/[*×][\d０１２３４５６７８９]+$/, '（すべて）'));

  const datalist = document.getElementById('list-of-core-part');
  datalist.innerHTML = '';

  if (items.length === 0) {
    return;
  }

  items.unshift("なし");

  items.forEach(
      item => {
        const option = document.createElement('option');
        option.textContent = item;
        datalist.appendChild(option);
      }
  );
}
// 戦利品欄 ----------------------------------------
// 追加
function addLoots(){
  let num = Number(form.lootsNum.value) + 1;
  let liNum = document.createElement('li');
  let liItem = document.createElement('li');
  liNum.id= idNumSet("loots-num");
  liItem.id= idNumSet("loots-item");
  liNum.innerHTML = '<span class="handle"></span><input type="text" name="loots'+num+'Num">';
  liItem.innerHTML = '<span class="handle"></span><input type="text" name="loots'+num+'Item">';
  document.getElementById("loots-num").appendChild(liNum);
  document.getElementById("loots-item").appendChild(liItem);
  
  form.lootsNum.value = num;
}
// 削除
function delLoots(){
  let num = Number(form.lootsNum.value);
  if(num > 1){
    if(form[`loots${num}Num`].value || form[`loots${num}Item`].value){
      if (!confirm(delConfirmText)) return false;
    }
    const listNum  = document.getElementById("loots-num");
    const listItem = document.getElementById("loots-item");
    listNum.lastElementChild.remove();
    listItem.lastElementChild.remove();
    num--;
    form.lootsNum.value = num;
  }
}
// ソート
setSortable('loots','#loots-num');
setSortable('loots','#loots-item');

File: _core/lib/sw2/edit-mons.pl
############# フォーム・モンスター #############
use strict;
#use warnings;
use utf8;
use open ":utf8";
use feature 'signatures';
no warnings 'experimental::signatures';

my $LOGIN_ID = $::LOGIN_ID;

### 読込前処理 #######################################################################################
require $set::lib_palette_sub;
### 各種データライブラリ読み込み --------------------------------------------------
require $set::data_mons;

### データ読み込み ###################################################################################
my ($data, $file, $message) = loadSheetData();
our %pc = %{ $data };

our $isNewSheet = isNewSheet();

### 出力準備 #########################################################################################
$message = applyMessageName($message, $pc{characterName} || $pc{monsterName} || '無題');

### 初期設定 --------------------------------------------------
if($isNewSheet){
  $pc{author} = (getPlayerName($LOGIN_ID))[0];
  $pc{protect} ||= $LOGIN_ID ? 'account' : 'password';
}
if($::mode eq 'edit' || ($::mode eq 'convert' && $pc{ver})){
  %pc = data_update_mons(\%pc);
}
elsif($::mode eq 'blanksheet'){
  $pc{paletteUseBuff} = 1;
  $pc{partsManualInput} = 0;

  %pc = applyCustomizedInitialValues(\%pc, 'm');
}

## カラー
setDefaultColors(\%pc);

## その他
$pc{partsNum}  ||= 1;
$pc{statusNum} ||= 1;
$pc{lootsNum}  ||= 2;

my $status_text_input = $pc{statusTextInput} || $pc{mount} || 0;

### 改行処理 --------------------------------------------------
convertEscapedBrToNewlines(\%pc,
  qw/skills description chatPalette/,
);

### フォーム表示 #####################################################################################
print renderEditPageStart(
  title => (removeTags removeRuby unescapeTags ( $pc{characterName} || $pc{monsterName} )),
  systemId => ($::SW2_0 ? 'sw2.0' : 'sw2.5'),
);
print renderEditHeaderMenu(
  tabsHtml => <<~'HTML',
    <li onclick="sectionSelect('common');" class="sheet-main"><span>魔物</span><span>データ</span>
    <li onclick="sectionSelect('palette');" class="unit-setting"><span><span class="shorten">ユニット(</span>コマ<span class="shorten">)</span></span><span>設定</span>
  HTML
);
print qq|<aside class="message">$message</aside>| if $message;

print <<"HTML";
  <section id="section-common">
    @{[ renderProtectBlock() ]}
    @{[ renderVisibilityBlock() ]}
    <div class="box in-toc" id="group" data-content-title="分類・タグ">
      <dl>
        <dt>分類</dt>
        <dd>
          <div class="select-input">
            <select name="taxa" oninput="selectInputCheck(this,'その他')">
            @{[
              map { '<option '. ($pc{taxa} eq @$_[0] ? ' selected': '') .">@$_[0]</option>" } @data::taxa
            ]}
            @{[
              ($pc{taxa} && !grep { @$_[0] eq $pc{taxa} } @data::taxa) ? "<option selected>$pc{taxa}</option>" : ''
            ]}
            </select>
            <input type="text" name="taxaFree">
          </div>
        <dd>@{[ checkbox 'mount','騎獣','checkMount' ]}
        <dt>タグ
        <dd>@{[ input 'tags' ]}
      </dl>
    </div>

    <div class="box in-toc" id="name-form" data-content-title="名称・製作者">
      <div>
        <dl id="character-name">
          <dt>名称
          <dd>@{[ input 'monsterName','text',"setName",'id="sub-name"' ]}
        </dl>
        <dl id="aka">
          <dt>名前
          <dd>@{[ input 'characterName','text','setName','id="main-name" placeholder="※名前を持つ魔物のみ"' ]}
        </dl>
      </div>
      <dl id="player-name">
        <dt>製作者
        <dd>@{[ input 'author' ]}
      </dl>
    </div>

    <div class="box status in-toc" data-content-title="基本データ">
      <dl class="mount-only price">
        <dt>価格
        <dd>購入@{[ input 'price' ]}G
        <dd>レンタル@{[ input 'priceRental' ]}G
        <dd>部位再生@{[ input 'priceRegenerate' ]}G
      </dl>
      <dl class="mount-only">
        <dt>適正レベル
        <dd>@{[ input 'lvMin','number','checkMountLevel','min="0"' ]} ～ @{[ input 'lvMax','number','checkMountLevel','min="0"' ]}
      </dl>
      <dl>
        <dt><span class="mount-only">騎獣</span>レベル
        <dd>@{[ input 'lv','number','checkLevel','min="0"' ]}
        <dd class="mount-only small">※入力すると、閲覧画面では現在の騎獣レベルのステータスのみ表示されます
      </dl>
      <dl>
        <dt>知能
        <dd>@{[ input 'intellect','','','list="data-intellect"' ]}
      </dl>
      <dl>
        <dt>知覚
        <dd>@{[ input 'perception','','','list="data-perception"' ]}
      </dl>
      <dl class="monster-only">
        <dt>反応
        <dd>@{[ input 'disposition','','','list="data-disposition"' ]}
      </dl>
      <dl>
        <dt>穢れ
        <dd>@{[ input 'sin','number','','min="0"' ]}
      </dl>
      <dl>
        <dt>言語
        <dd>@{[ input 'language','','','list="data-language"' ]}
      </dl>
      <dl class="monster-only">
        <dt>生息地
        <dd>@{[ input 'habitat' ]}
      </dl>
      <dl class="monster-only">
        <dt>知名度／弱点値
        <dd>@{[ input 'reputation' ]}／@{[ input 'reputation+','','','list="list-of-reputation-plus"' ]}
      </dl>
      <dl>
        <dt>弱点
        <dd>@{[ input 'weakness','','','list="data-weakness"' ]}
      </dl>
      <dl class="monster-only">
        <dt>先制値
        <dd>@{[ input 'initiative' ]}
      </dl>
      <dl>
        <dt>移動速度<dd>@{[ input 'mobility' ]}
      </dl>
      <dl class="monster-only">
        <dt>生命抵抗力
        <dd>@{[ input 'vitResist',($status_text_input ? 'text':'number'),'calcVit' ]} <span class=" calc-only">(@{[ input 'vitResistFix','number','calcVitF' ]})</span>
      </dl>
      <dl class="monster-only">
        <dt>精神抵抗力
        <dd>@{[ input 'mndResist',($status_text_input ? 'text':'number'),'calcMnd' ]} <span class=" calc-only">(@{[ input 'mndResistFix','number','calcMndF' ]})</span>
      </dl>
    </div>
    <p class="monster-only">@{[ input "statusTextInput",'checkbox','statusTextInputToggle']}命中・回避・抵抗に数値以外を入力</p>
    <div class="box in-toc" data-content-title="攻撃方法・命中・打撃・回避・防護・ＨＰ・ＭＰ">
    <table id="status-table" class="status">
      <thead>
        <tr>
          <th class="lv mount-only">Lv
          <th class="handle">
          <th class="name">攻撃方法<span class="text-part">（部位）</span>
          <th class="acc">命中力
          <th class="atk">打撃点
          <th class="eva">回避力
          <th class="def">防護点
          <th class="hp">ＨＰ
          <th class="mp">ＭＰ
          <th class="vit mount-only">生命抵抗
          <th class="mnd mount-only">精神抵抗
          <th>
        </tr>
      <tbody id="status-tbody">
        @{[ map {
          my $num = $_;
          $pc{"status${num}Damage"} = '2d+' if $pc{"status${num}Damage"} eq '' && $::mode eq 'blanksheet';
          <<~"ROW";
          <tr id="status-row${num}">
            <th class="mount-only">
            <td class="handle">
            <td>@{[ input "status${num}Style",'text',"checkStyle(${num}); updatePartsAutomatically()" ]}
            <td>@{[ input "status${num}Accuracy",($status_text_input ? 'text':'number'),"calcAcc($num)" ]}<span class="monster-only calc-only"><br>(@{[ input "status${num}AccuracyFix",'number',"calcAccF($num)" ]})</span>
            <td>@{[ input "status${num}Damage" ]}
            <td>@{[ input "status${num}Evasion",($status_text_input ? 'text':'number'),"calcEva($num)" ]}<span class="monster-only calc-only"><br>(@{[ input "status${num}EvasionFix",'number',"calcEvaF($num)" ]})</span>
            <td>@{[ input "status${num}Defense" ]}
            <td>@{[ input "status${num}Hp" ]}
            <td>@{[ input "status${num}Mp" ]}
            <td class="mount-only">@{[ input "status${num}Vit" ]}
            <td class="mount-only">@{[ input "status${num}Mnd" ]}
            <td><span class="button" onclick="addStatus(${num});">複<br>製</span>
          ROW
        } 1 .. $pc{statusNum} ]}
        @{[ map {
          my $lv = $_;
          <<~"TBODY";
          <tbody class="mount-only" id="status-tbody${lv}" data-lv="${lv}">
          @{[
            map {
              my $num = $_;
              $pc{"status${num}Damage"} = '2d+' if $pc{"status${num}Damage"} eq '' && $::mode eq 'blanksheet';
              <<~"ROW";
              <tr id="status-row${num}-${lv}">
                <th>
                <td>
                <td class="name" data-style="${num}">$pc{"status${num}Style"}
                <td>@{[ input "status${num}-${lv}Accuracy",($status_text_input ? 'text':'number') ]}
                <td>@{[ input "status${num}-${lv}Damage" ]}
                <td>@{[ input "status${num}-${lv}Evasion",($status_text_input ? 'text':'number') ]}
                <td>@{[ input "status${num}-${lv}Defense" ]}
                <td>@{[ input "status${num}-${lv}Hp" ]}
                <td>@{[ input "status${num}-${lv}Mp" ]}
                <td>@{[ input "status${num}-${lv}Vit" ]}
                <td>@{[ input "status${num}-${lv}Mnd" ]}
                <td>
              ROW
            } 1 .. $pc{statusNum}
          ]}
          TBODY
        } 2 .. ($pc{lvMax}-$pc{lvMin}+1) ]}
    </table>
    @{[ renderAddDelButtons('status') ]}
    </div>
    <div class="box parts in-toc" data-content-title="部位数・コア部位">
      @{[ checkbox 'partsManualInput', '部位数と内訳を手動入力する', 'updatePartsAutomatically' ]}
      <dl><dt>部位数<dd>@{[ input 'partsNum','number','','min="1"' ]} (@{[ input 'parts','','updatePartList' ]}) </dl>
      <dl><dt>コア部位<dd>@{[ input 'coreParts','','','list="list-of-core-part"' ]}</dl>
      <datalist id="list-of-core-part"></datalist>
    </div>
    <div class="box">
      <h2 class="in-toc">特殊能力</h2>
      <textarea name="skills">$pc{skills}</textarea>
      <div class="annotate">
        <b>行頭に</b>特殊能力の分類マークなどを記述すると、そこから次の「改行」または「全角スペース」までを自動的に見出し化します。<br>
         2.0での分類マークでも構いません。また、入力簡易化の為に入力しやすい代替文字での入力も可能です。<br>
         以下に見出しとして変換される記号・文字列を一覧にしています。<br>
        部位見出し（●）：<code>●</code><br>
        常時型　　（<i class="s-icon passive"></i>）：<code>[常]</code><code>○</code> <code>◯</code> <code>〇</code><br>
        ${\ do {
          if($::SW2_0){
            <<~"HTML";
            主動作型　（<i class="s-icon major0"   ></i>）：<code>[主]</code><code>＞</code> <code>▶</code> <code>〆</code><br>
            補助動作型（<i class="s-icon minor0"   ></i>）：<code>[補]</code><code>≫</code> <code>&gt;&gt;</code> <code>☆</code><br>
            宣言型　　（<i class="s-icon active0"  ></i>）：<code>[宣]</code><code>🗨</code> <code>□</code> <code>☑</code><br>
            条件型　　（<i class="s-icon condition"></i>）：<code>[条]</code><code>▽</code><br>
            条件選択型（<i class="s-icon selection"></i>）：<code>[選]</code><code>▼</code><br>
            HTML
          } else {
            <<~"HTML";
            戦闘準備型（<i class="s-icon setup"  ></i>）：<code>[準]</code><code>△</code><br>
            主動作型　（<i class="s-icon major"  ></i>）：<code>[主]</code><code>＞</code> <code>▶</code> <code>〆</code><br>
            補助動作型（<i class="s-icon minor"  ></i>）：<code>[補]</code><code>≫</code> <code>&gt;&gt;</code> <code>☆</code><br>
            宣言型　　（<i class="s-icon active" ></i>）：<code>[宣]</code><code>🗨</code> <code>□</code> <code>☑</code><br>
            HTML
          }
        } }
        <code>[]</code>で漢字一文字を囲う記法は、行頭でなくても各マークに変換されます。
      </div>
    </div>
    <div class="box loots monster-only">
      <h2 class="in-toc">戦利品</h2>
      <div id="loots-list">
        <ul id="loots-num">
          @{[ map {
            "<li id='loots-num${_}'><span class='handle'></span>".input("loots${_}Num",'','','list="data-roots-num"')
          } 1 .. $pc{lootsNum} ]}
        </ul>
        <ul id="loots-item">
          @{[ map {
            "<li id='loots-item${_}'><span class='handle'></span>".input("loots${_}Item")
          } 1 .. $pc{lootsNum} ]}
        </ul>
      </div>
      @{[ renderAddDelButtons('loots') ]}
    </div>
    <div class="box">
      <h2 class="in-toc">解説</h2>
      <textarea name="description">$pc{description}</textarea>
    </div>
  </section>
HTML
print renderChatPaletteForm();

print renderEditPageEnd(
  notes => '(C)Group SNE「ソード・ワールド'.($::SW2_0 ? '2.0' : '2.5').'」',
  multilineTargets => '「特殊能力」「解説」',
  addTextRule => renderAddTextRule(),
  extraHtml => renderDataList(),
);

sub renderDataList {
  return <<~"HTML";
  <datalist id="data-intellect">
    <option value="なし">
    <option value="動物並み">
    <option value="低い">
    <option value="人間並み">
    <option value="高い">
    <option value="命令を聞く">
  </datalist>
  <datalist id="data-perception">
    <option value="五感">
    <option value="五感（暗視）">
    <option value="五感（）">
    <option value="魔法">
    <option value="機械">
  </datalist>
  <datalist id="data-disposition">
    <option value="友好的">
    <option value="中立">
    <option value="敵対的">
    <option value="腹具合による">
    <option value="命令による">
  </datalist>
  <datalist id="data-language">
    <option value="なし">
  </datalist>
  <datalist id="list-of-reputation-plus">
    <option>―</option>
  </datalist>
  <datalist id="data-weakness">
    <option value="命中力+1">
    <option value="物理ダメージ+2点">
    <option value="魔法ダメージ+2点">
    <option value="属性ダメージ+3点">
    <option value="回復効果ダメージ+3点">
    <option value="なし">
  </datalist>
  <datalist id="data-roots-num">
    <option value="自動">
  </datalist>
  HTML
}

1;
